### PR TITLE
[processing] Port some stuff to c++

### PR DESCRIPTION
**Pull request by nyalldawson**
This ports parts of the provider and alglist processing classes to c++. As per the commit logs:

c++ base class for providers
-------------------------------------
Nothing particularly exciting here yet, but this commit moves the definition of the provider base class to a c++ QgsProcessingProvider abstract base class.

As part of this some existing python methods were renamed to make their use clearer and to fit with the QGIS c++ api conventions:

- getName was renamed to id
- getDescription was renamed to name
- getIcon was renamed to icon

New c++ QgsProcessingRegistry class
----------------------------------------------------
Intended as a registry for the various processing components, including providers, algorithms and various parameters and outputs.

Currently handles only registration of providers, as a step toward removing processing's algList (this requires first porting the algorithm class to c++)

A QgsProcessingRegistry instance is attached to QgsApplication, and is retrieved by QgsApplication::processingRegistry()

### CODE REVIEW

**Comment by nirvn** [comment]
@nyalldawson , FYI, when updating the processing's modeler to draw vectors instead of bitmaps, I noticed we'll need a way for providers and algorithms to return the file path of the icon (to be able to render .SVGs to a QPicture object).

**Comment by nyalldawson** [comment]
> FYI, when updating the processing's modeler to draw vectors instead of bitmaps, I noticed we'll need a way for providers and algorithms to return the file path of the icon (to be able to render .SVGs to a QPicture object).

@nirvn done, for those providers with a svg icon. Missing are saga, otb and lastools. The existing provider svgs could do with some tweaking too IMO to better work at the small icon sizes (eg gdal/grass icons are very busy)

**Comment by nirvn** [comment]
@nyalldawson , 👍 . The same svgIconPath() function will be needed when porting Algorithm to QgsProcessingAlgorithm.

You are right re existing SVGs in need of tweaking. That's on my to-do list.

**Comment by alexbruy** [comment]
Cool stuff.

File: python/core/processing/qgsprocessingprovider.sip [hunk 1]
```diff
@@ -0,0 +1,60 @@
...
+
+    /**
+     * Returns the full provider name, which is used to describe the provider within the GUI.
+     * This string should be localised.
+     * @see id()
+     */
+    virtual QString name() const = 0;
+
+    /**
+     * Returns true if the provider can be actived, or false if it is cannot be actived (eg due to
```
Comment (DelazJ) [inline python/core/processing/qgsprocessingprovider.sip]: activ**at**ed
it ~~is~~ cannot
eg or e.g. ? (let us be en-US :-) )

**Comment by nyalldawson** [comment]
@nirvn ok, over to you for visual improvements ;)

**Pull request merged**